packages/firefly/src/AppRouter.tsx

```diff
@@ -0,0 +1,188 @@
+import { useIsMswStarted } from "@squide/msw";
+import { useIsRouteMatchProtected, useLogger, useRoutes, type Route } from "@squide/react-router";
+import { useAreModulesReady, useAreModulesRegistered } from "@squide/webpack-module-federation";
+import { useEffect, useMemo, useState, type ReactNode } from "react";
+import { Outlet, RouterProvider, createBrowserRouter, useLocation, type RouterProviderProps } from "react-router-dom";
+
+export type OnLoadPublicDataFunction = (signal: AbortSignal) => Promise<unknown>;
+
+export type onLoadProtectedDataFunction = (signal: AbortSignal) => Promise<unknown>;
+
+export type OnCompleteRegistrationFunction = () => Promise<unknown>;
+
+interface RootRouteProps {
+    fallback: ReactNode;
+    onLoadPublicData?: OnLoadPublicDataFunction;
+    onLoadProtectedData?: onLoadProtectedDataFunction;
+    onCompleteRegistration?: OnCompleteRegistrationFunction;
+    waitForMsw: boolean;
+    areModulesRegistered: boolean;
+    areModulesReady: boolean;
+}
+
+// Most of the bootstrapping logic has been moved to this component because AppRouter
+// cannot leverage "useLocation" since it must be used in a child component of "RouterProvider".
+export function RootRoute(props: RootRouteProps) {
+    const {
+        fallback,
+        onLoadPublicData,
+        onLoadProtectedData,
+        onCompleteRegistration,
+        waitForMsw,
+        areModulesRegistered,
+        areModulesReady
+    } = props;
+
+    // Initialize as loaded if no handler is provided to load the public data.
+    const [isPublicDataLoaded, setIsPublicDataLoaded] = useState(!onLoadPublicData);
+
+    // Initialize as loaded if no handler is provided to load the protected data.
+    const [isProtectedDataLoaded, setIsProtectedDataLoaded] = useState(!onLoadProtectedData);
+
+    const logger = useLogger();
+    const location = useLocation();
+
+    // Re-render the app once MSW is started, otherwise, the API calls for module routes will return a 404 status.
+    const isMswStarted = useIsMswStarted(waitForMsw);
+
+    useEffect(() => {
+        // Only log these messages if MSW is enabled.
+        if (waitForMsw) {
+            if ((areModulesRegistered || areModulesReady) && !isMswStarted) {
+                logger.debug(`[shell] Modules are ${areModulesReady ? "ready" : "registered"}, waiting for MSW to start...`);
+            }
+
+            if (!areModulesRegistered && !areModulesReady && isMswStarted) {
```

Comment:
This should be an else if, as it's impossible that both conditions are met.

The first one has && !isMswStarted and the second && isMswStarted

Reply:
Fixed!